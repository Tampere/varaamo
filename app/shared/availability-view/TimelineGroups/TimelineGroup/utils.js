import { slotSize, slotWidth, slotMargin } from 'constants/SlotConstants';

import some from 'lodash/some';
import moment from 'moment';

function getTimeSlotWidth({ startTime, endTime } = {}) {
  const diff = endTime ? endTime.diff(startTime, 'minutes') : slotSize;
  const slots = Math.floor(diff / slotSize);

  return (slotWidth * slots) - slotMargin;
}

function getTimelineItems(date, reservations, resourceId) {
  const items = [];
  let reservationPointer = 0;
  let timePointer = date.clone().startOf('day');
  const end = date.clone().endOf('day');
  while (timePointer.isBefore(end)) {
    const reservation = reservations && reservations[reservationPointer];
    const isSlotReservation = reservation && timePointer.isSame(reservation.begin);
    if (isSlotReservation) {
      items.push({
        key: String(items.length),
        type: 'reservation',
        data: reservation,
      });
      timePointer = moment(reservation.end);
      reservationPointer += 1;
    } else {
      items.push({
        key: String(items.length),
        type: 'reservation-slot',
        data: {
          begin: timePointer.format(),
          end: timePointer.clone().add(slotSize, 'minutes').format(),
          resourceId,
          // isSelectable: false by default to improve selector performance by allowing
          // addSelectionData to make some assumptions.
          isSelectable: false,
        },
      });
      timePointer.add(slotSize, 'minutes');
    }
  }
  return items;
}

function isInsideOpeningHours(item, openingHours) {
  return some(openingHours, opening => (
    opening.opens <= item.data.begin && item.data.end <= opening.closes
  ));
}

function markItemSelectable(item, isSelectable, openingHours) {
  const selectable = (
    isSelectable
    && moment().isSameOrBefore(item.data.end)
    && (!openingHours || isInsideOpeningHours(item, openingHours))
  );
  return { ...item, data: { ...item.data, isSelectable: selectable } };
}

function markItemsSelectable(items, isSelectable, openingHours) {
  return items.map((item) => {
    if (item.type === 'reservation') return item;
    return markItemSelectable(item, isSelectable, openingHours);
  });
}

function addSelectionData(selection, resource, items) {
  if (!selection) {
    return markItemsSelectable(items, true, resource.openingHours);
  } if (selection.resourceId !== resource.id) {
    // isSelectable is false by default, so nothing needs to be done.
    // This is a pretty important performance optimization when there are tons of
    // resources in the AvailabilityView and the selection is in a state where the
    // first click has been done but the second (end time) hasn't. Without this
    // optimization we'd be calling markItemSelectable for every slot in every
    // resource when the user hovers to another slot.
    return items;
  }
  let lastSelectableFound = false;
  return items.map((item) => {
    if (lastSelectableFound || item.data.begin < selection.begin) {
      if (item.type === 'reservation') return item;
      // isSelectable is false by default.
      return item;
    }
    if (item.type === 'reservation') {
      lastSelectableFound = true;
      return item;
    }
    return markItemSelectable(item, true, resource.openingHours);
  });
}

export default {
  addSelectionData,
  getTimelineItems,
  getTimeSlotWidth,
};
