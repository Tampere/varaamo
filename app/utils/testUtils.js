import { shallow } from 'enzyme';
import forEach from 'lodash/forEach';
import forIn from 'lodash/forIn';
import get from 'lodash/get';
import mapValues from 'lodash/mapValues';
import merge from 'lodash/merge';
import set from 'lodash/set';
import React from 'react';
import { IntlProvider } from 'react-intl';
import { RSAA } from 'redux-api-middleware';

import rootReducer from 'state/rootReducer';
import enMessages from 'i18n/messages/en.json';

const testMessages = mapValues(enMessages, (value, key) => key);

function createApiTest(options) {
  /** Create some tests for an API action.
   *
   * Options can be an object like this:
   * {
   *  name: name of test (string),
   *  action: action creator (function),
   *  args: arguments for the action creator (array),
   *  tests: {
   *    method: HTTP method (string),
   *    endpoint: URL (string),
   *    headers: request headers (object)
   *    body: request body (string)
   *    meta: meta of the action (object)
   *    request: {
   *      type: request action type (string),
   *      hasSideEffect: (boolean)
   *      extra: {
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *    },
   *    success: {
   *      type: success action type (string),
   *      hasSideEffect: (boolean)
   *      payload: {
   *        data: payload response data (json-like)
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *      extra: {
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *    },
   *    error: {
   *      type: error action type (string),
   *      hasSideEffect: (boolean),
   *      extra: {
   *        tests: {
   *          (testName): ({ payload, promise, response }, (done)) (function)
   *        }
   *      }
   *    },
   *  },
   * }
   */
  const tests = options.tests;
  describe(options.name, () => {
    let action;
    let callAPI;

    beforeAll(() => {
      action = options.action.apply(null, options.args);
      callAPI = action[RSAA];
    });

    test('returns an object with a RSAA key', () => {
      expect(callAPI).toBeDefined();
    });

    tests.method && test(`uses ${tests.method}`, () => {
      expect(callAPI.method).toBe(tests.method);
    });

    tests.body && test('has correct body', () => {
      const body = (
        tests.body instanceof Object
          ? JSON.parse(callAPI.body)
          : callAPI.body
      );
      expect(body).toEqual(tests.body);
    });

    tests.endpoint && test('uses correct endpoint', () => {
      expect(callAPI.endpoint).toBe(tests.endpoint);
    });

    tests.meta && test('has correct meta', () => {
      expect(action.meta).toEqual(tests.meta);
    });

    describe('types', () => {
      const mockAction = {
        [RSAA]: {
          types: [{
            type: (
              (tests.request && tests.request.type) || 'Specify request.type'
            )
          }],
        },
      };

      ['request', 'success', 'error'].forEach((actionTypeName, index) => {
        tests[actionTypeName] && describe(actionTypeName, () => {
          const actionTests = tests[actionTypeName];
          const extraTests = actionTests.extra && actionTests.extra.tests;
          let typeAction;
          beforeAll(() => {
            typeAction = callAPI.types[index];
          });

          actionTests.type && test('has correct type', () => {
            expect(typeAction.type).toBe(actionTests.type);
          });

          extraTests && forIn(extraTests, (value, name) => {
            const getParams = () => ({
              typeAction,
              mockAction,
              meta: (
                actionTypeName === 'request'
                  ? typeAction.meta
                  : typeAction.meta(mockAction)
              ),
            });
            const func = (
              value.length === 2
                ? done => value(getParams(), done)
                : () => value(getParams())
            );
            test(name, func);
          });

          actionTypeName === 'success' && describe('payload', () => {
            let payload;
            let response;
            let promise;
            const payloadTests = actionTests.payload || {};
            const jsonData = payloadTests.data || {};

            beforeAll(() => {
              payload = typeAction.payload;
              response = {
                headers: {
                  get: () => 'application/json',
                },
                json: () => Promise.resolve(jsonData),
              };
            });

            beforeEach(() => {
              promise = payload({ type: actionTests.type }, {}, response);
            });

            test('exists', () => {
              expect(payload).toBeDefined();
            });

            payloadTests.tests && forIn(payloadTests.tests, (value, name) => {
              const func = (
                value.length === 2
                  ? done => value({ promise, payload, response }, done)
                  : () => value({ promise, payload, response })
              );
              test(name, func);
            });
          });
        });
      });
    });
  });
}

function getDefaultRouterProps() {
  return {
    location: {
      query: {},
    },
    params: {},
  };
}

function getInitialState() {
  return rootReducer(undefined, { type: 'NOOP' });
}

function getState(extraState = {}) {
  const state = getInitialState();
  const newState = {};

  forEach(Object.keys(extraState), (key) => {
    const mergedValue = merge({}, get(state, key), extraState[key]);
    set(newState, key, mergedValue);
  });
  return merge({}, state, newState);
}

function makeButtonTests(button, name, expectedText, expectedOnClickFunction) {
  test(`is an ${name} button`, () => {
    expect(button.props().children).toBe(expectedText);
  });

  test('clicking the button calls correct onClick function', () => {
    expectedOnClickFunction.reset();
    button.props().onClick();

    expect(expectedOnClickFunction.callCount).toBe(1);
  });
}

// Utilities for shallow testing components that use react-intl and injectT.
const intlProvider = new IntlProvider({ locale: 'en', messages: testMessages }, {});
const { intl } = intlProvider.getChildContext();

function shallowWithIntl(node, context) {
  const nodeWithIntl = React.cloneElement(node, { intl });
  return shallow(nodeWithIntl, { context: { ...context, intl } }).shallow({ context });
}

export {
  createApiTest,
  getDefaultRouterProps,
  getInitialState,
  getState,
  makeButtonTests,
  shallowWithIntl,
};
