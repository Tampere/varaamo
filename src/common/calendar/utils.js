import moment from 'moment';
import get from 'lodash/get';

/**
 * getDefaultSelectedTimeRange()
 * Get default selected time range from reservation data.
 * @returns {Object} selected time range
 */
export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end,
  } : null;
};


/**
 * isTimeRangeOverMaxPeriod();
 * Reservation cannot be longer than the resources max period if max period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than max_period or not
 */
export const isTimeRangeOverMaxPeriod = (resource, start, end, isStaff = false) => {
  const period = get(resource, 'max_period', null);

  if (!isStaff && period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') > durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * isTimeRangeUnderMinPeriod();
 * Reservation must longer than the resources min period if min period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than min_period or not
 */
export const isTimeRangeUnderMinPeriod = (resource, start, end) => {
  const period = get(resource, 'min_period', null);

  if (period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') < durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * addPeriodToTime()
 * Add period to input time
 * For example: period 00:30:00, time: moment(12:00:00) -> moment(12:30:00)
 *
 * @param {Moment} time
 * @param {string} period
 * @returns {Moment}
 */

const addPeriodToTime = (time, period) => {
  if (period) {
    const periodInMinute = moment.duration(period).as('minutes');
    return time.add(periodInMinute, 'minutes');
  }

  // Return original if no period
  return time;
};

/**
 * getMinPeriodTimeRange
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMinPeriodTimeRange = (resource, start, end) => {
  const minPeriod = get(resource, 'min_period', null);

  if (!minPeriod) {
    return {
      start, end,
    };
  }

  const minPeriodEnd = addPeriodToTime(moment(start), minPeriod);

  if (minPeriodEnd.diff(moment(end)) > 0) {
    // time range smaller than min_period, return min_period
    return {
      start,
      end: minPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};

/**
 * getMaxPeriodEndTime
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMaxPeriodTimeRange = (resource, start, end, isStaff) => {
  const maxPeriod = get(resource, 'max_period', null);

  if (!maxPeriod || isStaff) {
    return {
      start, end,
    };
  }

  const maxPeriodEnd = addPeriodToTime(moment(start), maxPeriod);

  if (moment(end).diff(maxPeriodEnd) > 0) {
    // time range is over max_period not allowed, return max_period
    return {
      start,
      end: maxPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};
