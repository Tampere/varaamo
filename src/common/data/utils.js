import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';
import camelCase from 'lodash/camelCase';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};

/**
 * Trim empty values from data object
 *
 * @param {Object} data
 * @returns {Object} Trimmed empty values object
 */

export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};

export const toCamelCase = (obj) => {
  if (!obj) return {};

  const camelCasedObj = {};

  Object.keys(obj).forEach((key) => {
    const camelCasedKey = camelCase(key);
    camelCasedObj[camelCasedKey] = obj[key];
  });

  return camelCasedObj;
};
