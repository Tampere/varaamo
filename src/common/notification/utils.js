import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';
/**
 * createNotification()
 * Creating notification popups
 *
 * @param {string} type
 * @param {string} [message='']
 * @param {string} [title='']
 * @param {number} [fadeTime=2000]
 * @param {function} callback
 * @returns {void}
 */
export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
