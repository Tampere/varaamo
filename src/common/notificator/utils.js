import moment from 'moment';

/**
 * createSelectOptions()
 * Create option arrays for selects.
 * @returns {Object} with select option arrays
 */
export const createSelectOptions = () => {
  const targetOptions = [
    { value: 'staff', label: 'Staff' },
    { value: 'user', label: 'User' },
    { value: 'all', label: 'All' },
  ];
  const urgencyOptions = [
    { value: 'common', label: 'Common' },
    { value: 'warning', label: 'Warning' },
    { value: 'danger', label: 'Danger' },
  ];
  const activeOptions = [
    { value: true, label: 'True' },
    { value: false, label: 'False' },
  ];

  return {
    targetOptions,
    urgencyOptions,
    activeOptions,
  };
};


/**
 * formatValuesForUse()
 * Format notification values to usable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForUse = (notification) => {
  const { targetOptions, urgencyOptions, activeOptions } = createSelectOptions();
  const formattedNotification = { ...notification };
  formattedNotification.target = targetOptions.find(option => option.value === notification.target);
  formattedNotification.urgency = urgencyOptions.find(option => option.value === notification.urgency);
  formattedNotification.active = activeOptions.find(option => option.value === notification.active);
  formattedNotification.until = moment(notification.until).toDate();
  return formattedNotification;
};

/**
 * formatValuesForDatabase()
 * Format notification values to savable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForDatabase = (notification) => {
  const formattedNotification = { ...notification };
  formattedNotification.target = notification.target.value;
  formattedNotification.urgency = notification.urgency.value;
  formattedNotification.active = notification.active.value;
  formattedNotification.until = moment(notification.until).format('YYYYMMDDTHHmmss');
  return formattedNotification;
};
