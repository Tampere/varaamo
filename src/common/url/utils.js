/**
 * Builds search string from an object.
 * @param search {object}
 * @returns {string}
 */
export const getSearch = (search) => {
  return new URLSearchParams(search).toString();
};

/**
 * Getter for link string with given path and query params.
 * @param path {string}
 * @param search {object}
 * @returns {string}
 */
export const getLinkString = (path, search = null) => {
  let searchString = '';

  if (search) {
    searchString = getSearch(search);
  }

  return `${path}${search ? `?${searchString}` : ''}`;
};
