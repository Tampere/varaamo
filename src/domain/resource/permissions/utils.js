import difference from 'lodash/difference';
import get from 'lodash/get';

import {
  resourceRoles, resourcePermissionTypes, resourcePermissionsByRole, UI_UNIT_STAFF_ROLES,
} from './constants';

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (!resourceRole) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}

const getUserPermissions = resource => (
  get(resource, 'user_permissions', false)
  || get(resource, 'userPermissions', undefined)
);

// Returns the role with the most permissions or null.
export function getUnitRoleFromResource(resource) {
  const userPermissions = getUserPermissions(resource);

  if (!userPermissions) {
    return null;
  }

  if (userPermissions.is_admin || userPermissions.isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (userPermissions.is_manager || userPermissions.isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (userPermissions.is_viewer || userPermissions.isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

export function getIsUnitStaff(unitRole) {
  return UI_UNIT_STAFF_ROLES.includes(unitRole);
}
