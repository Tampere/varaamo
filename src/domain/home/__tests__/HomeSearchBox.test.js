import React from 'react';
import simple from 'simple-mock';
import toJson from 'enzyme-to-json';

import HomeSearchBox from '../HomeSearchBox';
import { shallowWithIntl } from '../../../../app/utils/testUtils';

function getWrapper(props) {
  const defaults = {
    onChange: () => null,
    onSearch: () => null,
    value: 'meeting room',
  };
  return shallowWithIntl(<HomeSearchBox {...defaults} {...props} />);
}

describe('pages/home/HomeSearchBox', () => {
  test('render normally', () => {
    const wrapper = getWrapper();

    expect(toJson(wrapper)).toMatchSnapshot();
  });
  describe('handleSubmit', () => {
    test('calls props.onSearch', () => {
      const mockEvent = { preventDefault: () => null };
      const onSearch = simple.mock();
      const instance = getWrapper({ onSearch }).instance();
      instance.handleSubmit(mockEvent);
      expect(onSearch.callCount).toBe(1);
    });

    test('calls event.preventDefault', () => {
      const mockEvent = { preventDefault: simple.mock() };
      const instance = getWrapper().instance();
      instance.handleSubmit(mockEvent);
      expect(mockEvent.preventDefault.callCount).toBe(1);
    });
  });

  describe('handleChange', () => {
    test('calls this.props.onChange with correct value', () => {
      const instance = getWrapper().instance();
      const mockEvent = { preventDefault: simple.mock(), target: { value: 'some value' } };
      instance.handleChange(mockEvent);
      expect(instance.state.value).toBe(mockEvent.target.value);
    });
  });
});
