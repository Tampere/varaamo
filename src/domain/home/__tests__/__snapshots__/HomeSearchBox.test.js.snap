// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/home/HomeSearchBox render normally 1`] = `
<form
  className="app-HomeSearchBox"
  onSubmit={[Function]}
>
  <FormControl
    bsClass="form-control"
    className="app-HomeSearchBox__textfield"
    componentClass="input"
    onChange={[Function]}
    placeholder="HomeSearchBox.searchPlaceholder"
    type="text"
  />
  <Button
    active={false}
    block={false}
    bsClass="btn"
    bsStyle="primary"
    className="app-HomeSearchBox__button"
    disabled={false}
    type="submit"
  >
    HomeSearchBox.buttonText
  </Button>
</form>
`;
